trunk/src/app/srs_app_heartbeat.cpp

```diff
@@ -48,25 +50,84 @@ srs_error_t SrsHttpHeartbeat::do_heartbeat()
     std::string device_id = _srs_config->get_heartbeat_device_id();
-    
-    vector<SrsIPAddress*>& ips = srs_get_local_ips();
-    if (!ips.empty()) {
-        ip = ips[_srs_config->get_stats_network() % (int)ips.size()];
+
+    // Try to load the ip from the environment variable.
+    ip = srs_getenv("srs.device.ip"); // SRS_DEVICE_IP
+    if (ip.empty()) {
+        // Use the local ip address specified by the stats.network config.
+        vector<SrsIPAddress*>& ips = srs_get_local_ips();
+        if (!ips.empty()) {
+            ip = ips[_srs_config->get_stats_network() % (int) ips.size()]->ip;
+        }
     }
 
     SrsUniquePtr<SrsJsonObject> obj(SrsJsonAny::object());
 
     obj->set("device_id", SrsJsonAny::str(device_id.c_str()));
-    obj->set("ip", SrsJsonAny::str(ip->ip.c_str()));
+    obj->set("ip", SrsJsonAny::str(ip.c_str()));
+
+    SrsStatistic* stat = SrsStatistic::instance();
+    obj->set("server", SrsJsonAny::str(stat->server_id().c_str()));
+    obj->set("service", SrsJsonAny::str(stat->service_id().c_str()));
+    obj->set("pid", SrsJsonAny::str(stat->service_pid().c_str()));
     
     if (_srs_config->get_heartbeat_summaries()) {
         SrsJsonObject* summaries = SrsJsonAny::object();
         obj->set("summaries", summaries);
         
         srs_api_dump_summaries(summaries);
     }
+
+    if (_srs_config->get_heartbeat_ports()) {
+        // For RTMP listen endpoints.
+        if (true) {
+            SrsJsonArray* o = SrsJsonAny::array();
+            obj->set("rtmp", o);
+
+            vector<string> endpoints = _srs_config->get_listens();
+            for (int i = 0; i < (int) endpoints.size(); i++) {
+                o->append(SrsJsonAny::str(endpoints.at(i).c_str()));
+            }
+        }
+
+        // For HTTP Stream listen endpoints.
+        if (_srs_config->get_http_stream_enabled()) {
+            SrsJsonArray* o = SrsJsonAny::array();
+            obj->set("http", o);
+
+            string endpoint = _srs_config->get_http_stream_listen();
+            o->append(SrsJsonAny::str(endpoint.c_str()));
+        }
+
+        // For HTTP API listen endpoints.
+        if (_srs_config->get_http_api_enabled()) {
+            SrsJsonArray* o = SrsJsonAny::array();
+            obj->set("api", o);
+
+            string endpoint = _srs_config->get_http_api_listen();
+            o->append(SrsJsonAny::str(endpoint.c_str()));
+        }
+
+        // For SRT listen endpoints.
+        if (_srs_config->get_srt_enabled()) {
+            SrsJsonArray* o = SrsJsonAny::array();
+            obj->set("srt", o);
+
+            uint16_t endpoint = _srs_config->get_srt_listen_port();
+            o->append(SrsJsonAny::str(srs_int2str(endpoint).c_str()));
+        }
+
+        // For WebRTC listen endpoints.
+        if (_srs_config->get_rtc_server_enabled()) {
+            SrsJsonArray* o = SrsJsonAny::array();
+            obj->set("rtc", o);
+
+            int endpoint = _srs_config->get_rtc_server_listen();
```

Comment:
the `proxy` seems only support RTC over UDP right now, I guess `RTC over TCP` can be supported soon, so why not add tcp endpoint, `SrsConfig::get_rtc_server_tcp_listen`, before the new feature appear?

Reply:
Make sense. The UDP and TCP listen port is the same, so the rtc endpoint should be `udp://8000` or `tcp://8000` as such.